crates/env/src/engine/on_chain/impls.rs

```diff
@@ -573,38 +572,39 @@ impl TypedEnvBackend for EnvInstance {
             }
             Err(ext::Error::CalleeReverted) => {
-                let out =
-                    <ink_primitives::ConstructorResult<()> as scale::Decode>::decode(
-                        &mut &out_return_value[..],
-                    )
-                    .expect("Failed to decode into `ConstructorResult<()>`");
-
-                match out {
-                    ink_primitives::ConstructorResult::Ok(()) => {
-                        let out = <ink_primitives::ConstructorResult<
-                            core::result::Result<(), ContractError>,
-                        > as scale::Decode>::decode(
-                            &mut &out_return_value[..]
-                        )
-                        .expect("Failed to decode into a `ContractError`");
-
-                        match out {
-                            ink_primitives::ConstructorResult::Ok(contract_error) => {
-                                // This should only be an Err(ContractError), because if the call
-                                // to the constructor had succeeded we've be doing down the
-                                // "decoding into an `AccountId`" branch.
-                                //
-                                // Unwrapping the error is here to satisfy the type checker
-                                let contract_error = contract_error.unwrap_err();
-                                Ok(ink_primitives::ConstructorResult::Ok(Err(
-                                    contract_error,
-                                )))
-                            }
-                            _ => unreachable!("We already decoded into a `ConstructorResult::Ok` earlier."),
-                        }
+                let decoding_result = <ink_primitives::ConstructorResult<
+                    core::result::Result<(), ContractError>,
+                > as scale::Decode>::decode(
+                    &mut &out_return_value[..]
+                );
+
+                match decoding_result {
+                    Ok(constructor_result) => {
+                        let contract_result = constructor_result.expect(
+                            "If dispatch had failed, we shouldn't have been able to decode \
+                             the nested `Result`."
+                         );
+
+                        let contract_error = contract_result.expect_err(
+                            "Since the contract reverted, we only expect an `Error` from the constructor. \
+                             Otherwise we would be in the `AccountId` branch.");
+
+                        Ok(ink_primitives::ConstructorResult::Ok(Err(contract_error)))
                     }
-                    ink_primitives::ConstructorResult::Err(lang_err) => {
-                        Ok(ink_primitives::ConstructorResult::Err(lang_err))
+                    Err(_) => {
+                        // If we hit this branch it likely means dispatch failed, but we need to
```

Comment:
No, because in the branch above we would try and decode into a `Result<Result<()>>`, which as you mentioned earlier we'd fail to decode if we only had a `Result<()>` encoded into the buffer. In the piece of code you linked to we're encoding a `Result<(), LangError>` into the buffer.

This is also why we do another decode operation in this branch.

Reply:
In that case it is always the `Err` variant though, so it doesn't matter what type is specified as the `Ok` variant in the `return_value` generic parameters. It could be `()`, `Result<()>`, `Result<Result<Result<()>>`

Reply:
See https://github.com/paritytech/ink/pull/1532/files#diff-b240442c842f1df1ff5fdebd6d455289e5cfbbca404b8925e01ce4446db30a8eR128 for how I understand this test case. Correct me if I am still wrong about this.